pkg/engine/internal/scheduler/wire/proto_mapper.go

```diff
@@ -0,0 +1,570 @@
+	ts := &wirepb.TaskStatus{
+		State: m.taskStateToPbTaskState(from.State),
+	}
+
+	if from.Error != nil {
+		ts.Error = from.Error.Error()
+	}
+
+	return ts, nil
+}
+
+func (m *protoMapper) taskStateToPbTaskState(state workflow.TaskState) wirepb.TaskState {
+	switch state {
+	case workflow.TaskStateCreated:
+		return wirepb.TASK_STATE_CREATED
+	case workflow.TaskStatePending:
+		return wirepb.TASK_STATE_PENDING
+	case workflow.TaskStateRunning:
+		return wirepb.TASK_STATE_RUNNING
+	case workflow.TaskStateCompleted:
+		return wirepb.TASK_STATE_COMPLETED
+	case workflow.TaskStateCancelled:
+		return wirepb.TASK_STATE_CANCELLED
+	case workflow.TaskStateFailed:
+		return wirepb.TASK_STATE_FAILED
+	default:
+		return wirepb.TASK_STATE_INVALID
+	}
+}
+
+func (m *protoMapper) streamStateToPbStreamState(state workflow.StreamState) wirepb.StreamState {
+	switch state {
+	case workflow.StreamStateIdle:
+		return wirepb.STREAM_STATE_IDLE
+	case workflow.StreamStateOpen:
+		return wirepb.STREAM_STATE_OPEN
+	case workflow.StreamStateBlocked:
+		return wirepb.STREAM_STATE_BLOCKED
+	case workflow.StreamStateClosed:
+		return wirepb.STREAM_STATE_CLOSED
+	default:
+		return wirepb.STREAM_STATE_INVALID
+	}
+}
+
+func (m *protoMapper) nodeStreamMapToPbNodeStreamList(nodeMap map[physical.Node][]*workflow.Stream) (map[string]*wirepb.StreamList, error) {
+	result := make(map[string]*wirepb.StreamList)
+
+	for node, streams := range nodeMap {
+		// Get the node ID
+		nodeID := node.ID()
+		nodeIDStr := nodeID.String()
+
+		pbStreams := make([]*wirepb.Stream, len(streams))
+		for i, s := range streams {
+			pbStreams[i] = &wirepb.Stream{
+				Ulid:     protoUlid.ULID(s.ULID),
+				TenantId: s.TenantID,
+			}
+		}
+
+		result[nodeIDStr] = &wirepb.StreamList{
+			Streams: pbStreams,
+		}
+	}
+
+	return result, nil
+}
+
+// serializeArrowRecord serializes an Arrow record to bytes using IPC format.
+func (m *protoMapper) serializeArrowRecord(record arrow.Record) ([]byte, error) {
+	if record == nil {
+		return nil, errors.New("nil arrow record")
+	}
+
+	buf := &writerAdapter{}
+	writer := ipc.NewWriter(buf,
+		ipc.WithSchema(record.Schema()),
+		ipc.WithAllocator(m.allocator),
+	)
+	defer writer.Close()
+
+	if err := writer.Write(record); err != nil {
+		return nil, err
+	}
+
+	if err := writer.Close(); err != nil {
+		return nil, err
+	}
+
+	return buf.buf, nil
+}
+
+// writerAdapter adapts a byte buffer to io.Writer for Arrow IPC.
+type writerAdapter struct {
+	buf []byte
+}
+
+func (w *writerAdapter) Write(p []byte) (n int, err error) {
+	w.buf = append(w.buf, p...)
+	return len(p), nil
+}
+
+// deserializeArrowRecord deserializes an Arrow record from bytes using IPC format.
+func (m *protoMapper) deserializeArrowRecord(data []byte) (arrow.Record, error) {
+	if len(data) == 0 {
+		return nil, errors.New("empty arrow data")
+	}
+
+	reader, err := ipc.NewReader(
+		&readerAdapter{buf: data},
```

Comment:
nit: you can use `bytes.NewReader(data)` as the reader here, without needing to define readerAdapter